### DISCUSSION TITLE
eNVM Parameters (p_envm_params)

### DISCUSSION THREAD

**Discussion by Bharris100** [polarfire-soc/polarfire-soc-documentation#108] [HSS and U-Boot] [answered]
I am working on ymodem loader software that will allow my customers to complete HSS updates in the field.  The first function I am using is the uint8_t envm_init(const uint8_t * p_envm_params) function in mss_envm.c.  I have searched for proper settings for the following parameters:
    envmTrims.PREPGM_NDAC = (envmTrims.PREPGM_NDAC & 0x1f);
    envmTrims.PREPGM_PDAC = (envmTrims.PREPGM_PDAC & 0x1f);
    envmTrims.PGM_NDAC    = (envmTrims.PGM_NDAC & 0x1f);
    envmTrims.PGM_PDAC    = (envmTrims.PGM_PDAC & 0x1f);
    envmTrims.ERS_NDAC    = (envmTrims.ERS_NDAC & 0x1f);
    envmTrims.ERS_PDAC    = (envmTrims.ERS_PDAC & 0x1f);
    envmTrims.NEGMM_NDAC  = (envmTrims.NEGMM_NDAC & 0x1f);
    envmTrims.POSMM_PDAC  = (envmTrims.POSMM_PDAC & 0x1f);
Is there a resource that would include this information?
In addition, I am looking for software that will generate the .bin file for me from the .elf file.  Is that available somewhere as well?
Thank you in advance.

**Comment by hughbreslin** [comment, marked as answer]
Hi @Bharris100 

I've had a look and I can't see the parameters you're looking for unfortunately. The best reference I know of is in the system services, in the [Read eNVM Parameters Service](https://onlinedocs.microchip.com/pr/GUID-1409CF11-8EF9-4C24-A94E-70979A688632-en-US-1/index.html) but the advanced data sheet doesn't seem to have the required info. I'll let you know when I find out more. 

Re generating a bin from the elf file, could you use [objcopy](https://sourceware.org/binutils/docs/binutils/objcopy.html) to achieve this? Using `-O binary` should let you output a binary from the elf. If you're using SoftConsole you also have the option of outputting a binary file: 
![image](https://user-images.githubusercontent.com/44862881/176849946-4e6502bb-9697-44a3-a035-22a8edd5dd69.png)

**Reply by hughbreslin** [reply]
Also, if you're looking to do elf -> bin for the HSS, you should already have a bin in the "Default" folder where build files are output:
![image](https://user-images.githubusercontent.com/44862881/176850652-ff42cd89-bdb4-4e1b-b66d-9c4df933e95f.png)

You can also see an example of objcopy in the [HSS makefile](https://github.com/polarfire-soc/hart-software-services/blob/master/Makefile#L126) where it generates these files 🙂

**Reply by Bharris100** [reply]
@hughbreslin, Thanks for the information.  I saw much of this in the log file from my last use of SC to load the HSS.  One additional question, do I load the .bin file or the .hex file into the eNVM?  This is not clear in the log.  If I read the eNVM memory location I see that the data is similar to the .bin file, but the .bin file is byte swizzled, where the .hex file is not.
Thanks again.

**Reply by hughbreslin** [reply]
Is this through Libero? If its with your y-modem application I'm not sure, if its Libero you can use the boot mode 1 friendly hex file generated as Default/bootmode1/hss-envm-wrapper-bm1-p0.hex.

**Reply by Bharris100** [reply]
This is through the ymodem application.  I see the same file generated in SC under default/bootmode1.
From the log file I am seeing:
Invoking command: "C:\Microchip\SoftConsole-v2021.1\/riscv-unknown-elf-gcc/bin/riscv64-unknown-elf-objcopy.exe 
-I binary -O ihex --change-section-lma *+0x20220000 D:\HSS2021-11_Modified_rev02\hart-software-services\Default/bootmode1/hss-envm-wrapper-bm1-p0.bin 
D:\HSS2021-11_Modified_rev02\hart-software-services\Default/bootmode1/hss-envm-wrapper-bm1-p0.hex ".

**Reply by Bharris100** [reply]
Hi @hughbreslin,
Have you been able to find any information on the DAC parameters?
I think the answer to the which filetype to load, I have determined to be the hss-envm-wrapper-bm1-p0.bin file.
The interesting thing is that when I load a new .bin file using the functions in mss_envm.c using ymodem, I fail the verify portion right away.  If I use soft console and load the file and then load the same file using ymodem it says it passes, so I am coming to the conclusion that because the file is already in eNVM the verify works, so the program portion is not working in ymodem.  Stepping through the functions I am seeing what I expect in the correct data and addressing.
I have tried to read the address locations for the DAC parameters, and reading location 0x20200020 using read functions I wrote, I get a value that seems to be in all of the addresses around it making me doubt the result.  If I use the HSS hexdump function I am getting '0' when I read the same address.
Is the eNVM internal IP or 3rd party IP?  If it is 3rd party maybe I can get the parameters from that co.
Thanks

**Reply by hughbreslin** [reply]
Hey @Bharris100 unfortunately, I'm waiting to hear back from someone more in the know about this...

**Reply by Bharris100** [reply]
Hi @hughbreslin,
I stumbled on this little ditty in system services: MSS_SYS_read_envm_parameter(). This did exactly what I needed to do to complete my field updater.  This thread can be closed.  Thanks for your help.
